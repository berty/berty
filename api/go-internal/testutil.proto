syntax = "proto3";

package testutil;

option go_package = "berty.tech/berty/go/internal/testutil";

service TestService {
  // use to test unary
  rpc EchoTest(EchoTest.Request) returns (EchoTest.Reply);

  // Use to test stream.
  rpc EchoStreamTest(EchoStreamTest.Request) returns (stream EchoStreamTest.Reply);

  // Use to test duplex stream.
  rpc EchoDuplexTest(stream EchoDuplexTest.Request) returns (stream EchoDuplexTest.Reply);
}

message EchoTest {
  message Request {
    int64 delay = 1;
    string echo = 2;
    bool trigger_error = 3;
  }
  message Reply {
    string echo = 1;
  }
}

message EchoStreamTest {
  message Request {
    int64 delay = 1;
    string echo = 2;
    bool trigger_error = 3;
  }
  message Reply {
    string echo = 1;
  }
}

message EchoDuplexTest {
  message Request {
    string echo = 1;
    bool trigger_error = 2;
  }
  message Reply {
    string echo = 1;
  }
}
