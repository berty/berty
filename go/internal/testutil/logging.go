package testutil

import (
	"flag"
	"os"
	"strings"
	"sync"
	"testing"

	"github.com/stretchr/testify/assert"
	"go.uber.org/zap"

	"berty.tech/berty/v2/go/internal/logutil"
)

const defaultLoggingFilters = "info+:bty.test* error+:*,-ipfs*"

var (
	logFilters     = flag.String("log-filters", defaultLoggingFilters, "log namespaces")
	logFile        = flag.String("log-file", "", "log to file")
	logFormat      = flag.String("log-format", "color", "json, console, color")
	loggerInstance *zap.Logger
	loggerCleanup  func()
	loggerInitOnce sync.Once
)

func Logger(t *testing.T) (*zap.Logger, func()) {
	t.Helper()

	loggerInitOnce.Do(func() {
		if val := os.Getenv("BERTY_LOGFILTERS"); val != "" {
			*logFilters = val
		}
		if val := os.Getenv("BERTY_LOGFILE"); val != "" {
			*logFile = val
		}
		if val := os.Getenv("BERTY_LOGFORMAT"); val != "" {
			*logFormat = val
		}
		*logFilters = strings.ReplaceAll(*logFilters, ":default:", defaultLoggingFilters)

		var err error
		loggerInstance, loggerCleanup, err = logutil.NewLogger(*logFilters, *logFormat, *logFile)
		if !assert.NoError(t, err) {
			loggerInstance = zap.NewNop()
			loggerCleanup = func() {}
		}
	})

	return loggerInstance, loggerCleanup
}
