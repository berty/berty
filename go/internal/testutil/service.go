package testutil

import (
	context "context"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"go.uber.org/zap"
	"google.golang.org/grpc"

	"berty.tech/berty/v2/go/pkg/errcode"
	"berty.tech/weshnet/v2/pkg/grpcutil"
)

type service struct {
	logger *zap.Logger

	UnimplementedTestServiceServer
}

type Options struct {
	Logger *zap.Logger
}

func (o *Options) applyDefault() {
	if o.Logger == nil {
		o.Logger = zap.NewNop()
	}
}

func TestingNewServiceClient(ctx context.Context, t *testing.T, opts *Options) (*grpc.ClientConn, *grpc.Server) {
	t.Helper()

	srv := grpc.NewServer()
	svc := NewService(opts)
	RegisterTestServiceServer(srv, svc)

	l := grpcutil.NewBufListener(2048)

	cc, err := l.NewClientConn(ctx)
	assert.NoError(t, err)

	go func() {
		_ = srv.Serve(l.Listener)
	}()

	t.Cleanup(func() { l.Close() })

	return cc, srv
}

func NewService(opts *Options) TestServiceServer {
	opts.applyDefault()
	return &service{
		logger: opts.Logger,
	}
}

func (svc *service) EchoTest(_ context.Context, req *EchoTest_Request) (*EchoTest_Reply, error) {
	if req.TriggerError {
		return nil, errcode.ErrCode_ErrTestEcho
	}

	if req.Delay > 0 {
		time.Sleep(time.Duration(req.Delay) * time.Millisecond)
	}

	return &EchoTest_Reply{Echo: req.Echo}, nil
}

func (svc *service) EchoStreamTest(req *EchoStreamTest_Request, srv TestService_EchoStreamTestServer) error {
	if req.TriggerError {
		return errcode.ErrCode_ErrTestEcho
	}

	for {
		err := srv.Send(&EchoStreamTest_Reply{Echo: req.Echo})
		if err != nil {
			return errcode.ErrCode_ErrTestEchoSend.Wrap(err)
		}

		time.Sleep(time.Duration(req.Delay) * time.Millisecond)
	}
}

func (svc *service) EchoDuplexTest(srv TestService_EchoDuplexTestServer) error {
	for {
		req, err := srv.Recv()
		if err != nil {
			return errcode.ErrCode_ErrTestEchoRecv.Wrap(err)
		}

		if req.TriggerError {
			return errcode.ErrCode_ErrTestEcho
		}

		err = srv.Send(&EchoDuplexTest_Reply{
			Echo: req.Echo,
		})
		if err != nil {
			return errcode.ErrCode_ErrTestEchoSend.Wrap(err)
		}
	}
}
