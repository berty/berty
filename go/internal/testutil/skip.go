package testutil

import (
	"os"
	"testing"
)

func SkipSlow(t *testing.T) {
	t.Helper()
	if os.Getenv("SKIP_SLOW_TESTS") == "1" {
		t.Skip("slow test skipped")
	}
}
