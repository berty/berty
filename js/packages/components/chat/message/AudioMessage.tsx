import React, { useEffect, useMemo } from 'react'
import { TouchableOpacity, View } from 'react-native'
import { Icon, Text } from '@ui-kitten/components'
import moment from 'moment'

import { useMessengerContext, useThemeColor } from '@berty-tech/store'
import { useStyles } from '@berty-tech/styles'
import { EndError, PlayerItemMetadata, useMusicPlayer } from '@berty-tech/music-player'
import beapi from '@berty-tech/api'
import { playSoundAsync } from '@berty-tech/store/sounds'

import { limitIntensities, voiceMemoFilename } from '../record/common'
import { getSource } from '../../utils'

const volumeValueShown = 50

const normalizeVolumeIntensities = (intensities: Array<number>) => {
	const min = Math.min(...intensities)
	const max = Math.max(...intensities)

	intensities = limitIntensities(
		intensities.map(i => (i - min) / (max - min)),
		volumeValueShown,
	)

	return intensities
}

export const WaveForm: React.FC<{
	intensities: any[]
	duration: number | null
	currentTime?: number
}> = ({ intensities, duration, currentTime = 0 }) => {
	const normalizedIntensities = useMemo(
		() => normalizeVolumeIntensities(intensities),
		[intensities],
	)
	const [{ margin, text }] = useStyles()
	const colors = useThemeColor()
	return (
		<View
			style={{
				flexDirection: 'row',
				alignItems: 'center',
				flex: 1,
				padding: 8,
				height: '100%',
			}}
		>
			<View style={{ flex: 1, flexDirection: 'row', height: '100%', alignItems: 'center' }}>
				{normalizedIntensities.map((intensity, index) => {
					return (
						<React.Fragment key={index}>
							<View
								style={{
									backgroundColor:
										duration &&
										index < Math.floor((currentTime! / duration) * normalizedIntensities.length)
											? colors['main-background']
											: colors['background-header'],
									minHeight: 5,
									height: 70 * intensity + '%',
									flex: 2,
									borderRadius: 5,
								}}
							/>
							<View style={{ flex: 1 }} />
						</React.Fragment>
					)
				})}
			</View>
			<Text style={[{ color: colors['background-header'] }, margin.left.tiny, text.size.small]}>
				{moment.utc(duration).format('mm:ss')}
			</Text>
		</View>
	)
}

const AudioPreview: React.FC<{
	media: beapi.messenger.IMedia
	currentTime?: number
}> = ({ media, currentTime = 0 }) => {
	const colors = useThemeColor()
	const [normalizedIntensities, duration] = useMemo(() => {
		const metadata = beapi.messenger.MediaMetadata.decode(media.metadataBytes!)
		const previews = metadata?.items.filter(
			m => m.metadataType === beapi.messenger.MediaMetadataType.MetadataAudioPreview,
		)

		if (previews === undefined || previews.length === 0) {
			return [null, null]
		}

		const preview = beapi.messenger.AudioPreview.decode(previews[0].payload!)
		return [normalizeVolumeIntensities(preview.volumeIntensities), preview.durationMs]
	}, [media.metadataBytes])

	if (normalizedIntensities === null) {
		return (
			<View style={{ flex: 1 }}>
				<Text style={{ color: colors['background-header'] }} numberOfLines={1}>
					{media.filename!}
				</Text>
			</View>
		)
	}

	return (
		<WaveForm intensities={normalizedIntensities} duration={duration} currentTime={currentTime} />
	)
}

export const AudioMessage: React.FC<{
	medias: Array<beapi.messenger.IMedia>
	onLongPress: () => void
	isHighlight: boolean
}> = ({ medias, onLongPress, isHighlight }) => {
	const colors = useThemeColor()
	const { protocolClient, client } = useMessengerContext()
	const [{ padding, border, margin }, { windowWidth, scaleSize }] = useStyles()
	const { player: globalPlayer, load: globalPlayerLoad, handlePlayPause } = useMusicPlayer()
	const cid = useMemo(() => medias[0].cid, [medias])
	const filename = useMemo(() => medias[0].filename, [medias])

	const isPlaying = useMemo(
		() => globalPlayer.metadata?.id === cid && globalPlayer.player?.isPlaying === true,
		[globalPlayer.metadata?.id, globalPlayer.player?.isPlaying, cid],
	)

	useEffect(() => {
		if (!isPlaying) {
			return
		}

		return () => {
			if (filename !== voiceMemoFilename || globalPlayer.player?.isStopped === false) {
				return
			}

			globalPlayerLoad(
				(async (): Promise<[string, PlayerItemMetadata]> => {
					const next = await client?.mediaGetRelated({
						cid: cid,
						fileNames: [voiceMemoFilename],
					})

					if (!next || next.end) {
						throw new EndError()
					}

					const srcP = getSource(protocolClient!, next.media?.cid!)
					await playSoundAsync('contactRequestAccepted')
					const src = await srcP

					return [
						`data:${next.media?.mimeType};base64,${src}`,
						{
							id: next.media?.cid!,
						},
					]
				})(),
			)
		}
	}, [cid, isPlaying, globalPlayer.player, globalPlayerLoad, client, protocolClient, filename])

	return (
		<TouchableOpacity
			style={{ alignItems: 'center' }}
			onLongPress={onLongPress}
			activeOpacity={0.9}
		>
			<View
				style={[
					{
						backgroundColor: colors['input-background'],
						alignItems: 'center',
						justifyContent: 'center',
						height: 50,
						width: windowWidth - 100,
						maxWidth: 400,
						flexDirection: 'row',
					},
					border.radius.small,
					isHighlight && {
						borderColor: colors['background-header'],
						borderWidth: 1,
						shadowColor: colors.shadow,
						shadowOffset: {
							width: 0,
							height: 8,
						},
						shadowOpacity: 0.44,
						shadowRadius: 10.32,
						elevation: 16,
					},
				]}
			>
				<TouchableOpacity
					onPress={() => {
						if (!cid) {
							return
						}
						if (globalPlayer.metadata?.id === medias[0].cid) {
							handlePlayPause()
						} else if (protocolClient) {
							globalPlayerLoad(
								getSource(protocolClient, cid).then(src => [
									`data:${medias[0].mimeType};base64,${src}`,
									{
										id: cid,
									},
								]),
							)
						}
					}}
					style={[
						padding.top.tiny,
						padding.left.scale(10),
						border.radius.small,
						margin.tiny,
						{
							alignSelf: 'center',
							alignItems: 'center',
							justifyContent: 'center',
						},
					]}
				>
					<Icon
						name={isPlaying ? 'pause' : 'play'}
						fill={colors['background-header']}
						height={26 * scaleSize}
						width={26 * scaleSize}
						pack='custom'
					/>
				</TouchableOpacity>
				<AudioPreview
					media={medias[0]}
					currentTime={isPlaying ? globalPlayer.player?.currentTime : 0}
				/>
			</View>
		</TouchableOpacity>
	)
}
