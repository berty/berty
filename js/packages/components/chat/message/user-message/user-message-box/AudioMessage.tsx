import { Icon } from '@ui-kitten/components'
import React, { useMemo } from 'react'
import { View, TouchableOpacity, ActivityIndicator, StyleSheet } from 'react-native'
import { useSelector } from 'react-redux'

import beapi from '@berty/api'
import { useAppDimensions } from '@berty/contexts/app-dimensions.context'
import { useMusicPlayer } from '@berty/contexts/musicPlayer.context'
import { useStyles } from '@berty/contexts/styles'
import { useThemeColor } from '@berty/hooks'
import { selectProtocolClient } from '@berty/redux/reducers/ui.reducer'

import { AudioPreviewPriv } from '../../audio/AudioPreview.priv'
import { AudioWrapperPriv } from '../../audio/AudioWrapper.priv'

interface AudioMessageProps {
	medias: Array<beapi.messenger.IMedia>
	isHighlight: boolean
	onLongPress: () => void
	isMine: boolean
}

export const AudioMessage: React.FC<AudioMessageProps> = props => {
	const colors = useThemeColor()
	const protocolClient = useSelector(selectProtocolClient)
	const { padding, border, margin } = useStyles()
	const { windowWidth } = useAppDimensions()
	const {
		player: globalPlayer,
		load: globalPlayerLoad,
		handlePlayPause,
		loading: globalPlayerLoading,
		playing: globalPlayerPlaying,
		currentTime,
	} = useMusicPlayer()
	const mediaCID = useMemo(() => props.medias[0].cid, [props.medias])
	const mimeType = (props.medias.length > 0 && props.medias[0].mimeType) || 'blob'

	const isCurrent = globalPlayer.metadata?.id === mediaCID
	const isPlaying = isCurrent && globalPlayerPlaying
	const loading = isCurrent && globalPlayerLoading === true

	const onPress = () => {
		if (
			globalPlayer.metadata?.id === props.medias[0].cid &&
			globalPlayer.player?.currentTime !== -1
		) {
			handlePlayPause()
		} else if (protocolClient && mediaCID) {
			globalPlayerLoad(mediaCID, mimeType)
		}
	}

	return (
		<AudioWrapperPriv isMine={props.isMine} onLongPress={props.onLongPress}>
			<View
				style={[
					styles.content,
					{
						backgroundColor: colors['background-header'],
						width: windowWidth - 100,
					},
					border.radius.big,
					props.isHighlight && {
						...styles.highlightContent,
						borderColor: colors['background-header'],
						shadowColor: colors.shadow,
					},
				]}
			>
				<TouchableOpacity
					onPress={onPress}
					disabled={loading}
					style={[
						padding.left.scale(10),
						border.radius.small,
						margin.left.small,
						margin.right.tiny,
						styles.button,
					]}
				>
					<View
						style={[
							padding.scale(7),
							isCurrent ? border.radius.scale(10) : border.radius.scale(12),
							{
								backgroundColor: `${colors['main-background']}80`,
								borderWidth: 1,
								borderColor: '#0000',
							},
							isCurrent && {
								borderColor: `${colors['reverted-main-text']}50`,
							},
						]}
					>
						{loading ? (
							<ActivityIndicator color={colors['reverted-main-text']} size={16} />
						) : (
							<Icon
								name={isPlaying ? 'pause' : 'play'}
								fill={colors['reverted-main-text']}
								height={16}
								width={16}
								pack='custom'
							/>
						)}
					</View>
				</TouchableOpacity>
				<AudioPreviewPriv media={props.medias[0]} currentTime={isCurrent ? currentTime : 0} />
			</View>
		</AudioWrapperPriv>
	)
}

const styles = StyleSheet.create({
	content: {
		alignItems: 'center',
		justifyContent: 'center',
		height: 50,
		maxWidth: 400,
		flexDirection: 'row',
	},
	highlightContent: {
		borderWidth: 1,
		shadowOffset: {
			width: 0,
			height: 8,
		},
		shadowOpacity: 0.44,
		shadowRadius: 10.32,
		elevation: 16,
	},
	button: {
		alignSelf: 'center',
		alignItems: 'center',
		justifyContent: 'center',
	},
})
