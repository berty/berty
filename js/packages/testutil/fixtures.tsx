import beapi from '@berty-tech/api'

const empty: beapi.messenger.IMedia = {
	metadataBytes: beapi.messenger.MediaMetadata.encode({}).finish(),
}

const media = {
	empty,
}

export default {
	media,
}
