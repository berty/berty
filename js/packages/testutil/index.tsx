import React from 'react'
import { NavigationContainer } from '@react-navigation/native'
import { IconRegistry } from '@ui-kitten/components'
import { EvaIconsPack } from '@ui-kitten/eva-icons'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import { Provider as ThemeProvider } from '@berty-tech/components/theme'
import { FeatherIconsPack } from '@berty-tech/messenger-app/feather-icons'
import { CustomIconsPack } from '@berty-tech/messenger-app/custom-icons'
import { Provider as StyleProvider } from '@berty-tech/styles'

export const TestProvider: React.FC = ({ children }) => (
	<SafeAreaProvider
		initialMetrics={{
			frame: { x: 0, y: 0, width: 0, height: 0 },
			insets: { top: 0, left: 0, right: 0, bottom: 0 },
		}}
	>
		<StyleProvider>
			<IconRegistry icons={[EvaIconsPack, FeatherIconsPack, CustomIconsPack]} />
			<ThemeProvider>
				<NavigationContainer>{children}</NavigationContainer>
			</ThemeProvider>
		</StyleProvider>
	</SafeAreaProvider>
)

import _fixtures from './fixtures'
export const fixtures = _fixtures
