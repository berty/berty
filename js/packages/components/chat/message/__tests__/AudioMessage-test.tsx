import React from 'react'
import renderer from 'react-test-renderer'

import { fixtures, TestProvider } from '@berty-tech/testutil'

import { AudioMessage } from '../AudioMessage'

test('AudioMessage renders correctly', () => {
	const props: React.ComponentProps<typeof AudioMessage> = {
		medias: [fixtures.media.empty],
		onLongPress: () => {},
		isHighlight: false,
	}
	// FIXME: mock useContactRequest to get children in the snapshot
	const tree = renderer
		.create(
			<TestProvider>
				<AudioMessage {...props} />
			</TestProvider>,
		)
		.toJSON()
	expect(tree).toMatchSnapshot()
})
