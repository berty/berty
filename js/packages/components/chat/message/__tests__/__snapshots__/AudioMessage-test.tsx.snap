// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioMessage renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#F7F8FF",
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "center",
              "maxWidth": 400,
              "width": 650,
            },
            Object {
              "borderRadius": 9,
            },
            false,
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "borderRadius": 9,
              "justifyContent": "center",
              "margin": 4,
              "opacity": 1,
              "paddingLeft": 10,
              "paddingTop": 4,
            }
          }
        >
          <View
            style={
              Object {
                "transform": Array [
                  Object {
                    "scale": 1,
                  },
                ],
              }
            }
          >
            <SvgMock
              color="#525BEC"
              height={26}
              style={Array []}
              width={26}
            />
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#222B45",
                  "fontFamily": "Open Sans",
                  "fontSize": 15,
                  "fontWeight": "600",
                },
                Object {
                  "color": "#525BEC",
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
